[4.0] password field as custom element

dgrammatiko opened the pull request:
Pull Request for Issue # .

### Summary of Changes
Fields with custom javascript functionality **should** always be custom elements!!!

So, this PR is doing that and also:
- removes the dependency from Bootstrap
- removes the dependency from Font awesome
- Should be accessible
- Should be extremely easy to customise
- Proper implementation of custom elements (eg works asynchronously)


### Testing Instructions
Apply patch and check the
- Login page in the back end
- The login module in the front end
- The edit user page in the backend

### Preview
Backend login: 
![screenshot 2018-12-01 at 15 41 23](https://user-images.githubusercontent.com/3889375/49329471-a7d1db00-f57f-11e8-8bac-c283a0230eaa.png)

Frontend Login:
![screenshot 2018-12-01 at 15 41 43](https://user-images.githubusercontent.com/3889375/49329473-b0c2ac80-f57f-11e8-91e5-62d789b55c08.png)

Edit user page: 
![screenshot 2018-12-01 at 15 41 12](https://user-images.githubusercontent.com/3889375/49329478-c0da8c00-f57f-11e8-8602-89063d5e9acb.png)
 
### Expected result



### Actual result



### Documentation Changes Required


Some Notes here: 

The input prepend/append thingy from the Bootstrap is utterly crap for 2 reasons:
- if used without a proper design language (which of course such thing doesn't exist for Joomla, although I was screaming for it) then the UX is getting inconsistent but more importantly is becoming harder for people to familiarise with all the extra (unneeded) graphics.

- the markup is also crap, eg a span element cannot be focusable (unless if you start doing things like type=button, etc) neither does implements all the built in keyboard functionality. 

All and all: **do not use the input-group** markup/classes just for decoration purposes!!!!


For those really unfamiliar with design language and the necessity for such thing established for Joomla, please check https://design.trello.com/components/form or https://github.com/alexpate/awesome-design-systems

dgrammatiko committed:
init

dgrammatiko committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into 4.0-dev-CE-password-field

* '4.0-dev' of github.com:joomla/joomla-cms:
  [4.0] RFC Quickicons Joomla and Extensions updates (#23095)

dgrammatiko committed:
more

dgrammatiko committed:
cs

brianteeman commented on layouts/joomla/form/field/password.php:
```diff
@@ -50,53 +50,41 @@
 	strlen($hint) ? 'placeholder="' . htmlspecialchars($hint, ENT_COMPAT, 'UTF-8') . '"' : '',
-	!$autocomplete ? 'autocomplete="off"' : '',
-	!empty($class) ? 'class="form-control ' . $class . '"' : 'class="form-control"',
+	!$autocomplete ? 'autocomplete="passowrd"' : 'autocomplete="' . $autocomplete . '"',
```
passowrd - typo??

dgrammatiko committed:
cs

SharkyKZ commented on modules/mod_login/tmpl/default.php:
```diff
@@ -33,53 +33,42 @@
 			<?php else : ?>
 				<label for="modlgn-username"><?php echo Text::_('MOD_LOGIN_VALUE_USERNAME'); ?></label>
-				<input id="modlgn-username" type="text" name="username" class="form-control" placeholder="<?php echo Text::_('MOD_LOGIN_VALUE_USERNAME'); ?>">
+				<input class="form-control" id="modlgn-username" type="text" name="username" autocomplete="username" laceholder="<?php echo Text::_('MOD_LOGIN_VALUE_USERNAME'); ?>">
```
Typo
```suggestion
				<input class="form-control" id="modlgn-username" type="text" name="username" autocomplete="username" placeholder="<?php echo Text::_('MOD_LOGIN_VALUE_USERNAME'); ?>">
```

Commit:
Update modules/mod_login/tmpl/default.php

Co-Authored-By: dgrammatiko <d.grammatiko@gmail.com>

infograf768 commented:
Failure:
This breaks in multilingual site the automatic language change with the error:
`Call to undefined method Joomla\CMS\Language\Language::setLanguage() `

```

Call stack
--
# | Function | Location
1 | () | JROOT/plugins/system/languagefilter/languagefilter.php:726
2 | PlgSystemLanguageFilter->onUserLogin() | JROOT/libraries/src/Plugin/CMSPlugin.php:287
3 | Joomla\CMS\Plugin\CMSPlugin->Joomla\CMS\Plugin\{closure}() | JROOT/libraries/vendor/joomla/event/src/Dispatcher.php:400
4 | Joomla\Event\Dispatcher->dispatch() | JROOT/libraries/src/Application/EventAware.php:111
5 | Joomla\CMS\Application\WebApplication->triggerEvent() | JROOT/libraries/src/Application/CMSApplication.php:819
6 | Joomla\CMS\Application\CMSApplication->login() | JROOT/libraries/src/Application/SiteApplication.php:716
7 | Joomla\CMS\Application\SiteApplication->login() | JROOT/components/com_users/Controller/UserController.php:122
8 | Joomla\Component\Users\Site\Controller\UserController->login() | JROOT/libraries/src/MVC/Controller/BaseController.php:734
9 | Joomla\CMS\MVC\Controller\BaseController->execute() | JROOT/libraries/src/Dispatcher/ComponentDispatcher.php:146
10 | Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch() | JROOT/libraries/src/Component/ComponentHelper.php:380
11 | Joomla\CMS\Component\ComponentHelper::renderComponent() | JROOT/libraries/src/Application/SiteApplication.php:211
12 | Joomla\CMS\Application\SiteApplication->dispatch() | JROOT/libraries/src/Application/SiteApplication.php:250
13 | Joomla\CMS\Application\SiteApplication->doExecute() | JROOT/libraries/src/Application/CMSApplication.php:233
14 | Joomla\CMS\Application\CMSApplication->execute() | JROOT/includes/app.php:63
15 | require_once() | JROOT/index.php:36

```

infograf768 commented:
Looks like this happens only when using the module, not a login menu item.

infograf768 commented on build/media/webcomponents/js/field-password/field-password.js:
```diff
@@ -0,0 +1,288 @@
+ * License: MIT
+ * Repo: https://github.com/tkjaergaard/Password-Strength
+ *
+ * Adopted by Dimitris Grammatikogiannis
```
`Adapted` I guess.

dgrammatiko replied:
oops

infograf768 commented:
I have tested this item :white_check_mark: successfully on e293a43311e0db47dfd579b0603f4f9203bac547<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23215">issues.joomla.org/tracker/joomla-cms/23215</a>.</sub>

dgrammatiko committed:
Adopting scripts is not a good idea

Fedik commented on build/media/webcomponents/js/field-password/field-password.js:
```diff
@@ -0,0 +1,288 @@
+    this.isVisible = false;
+
+    // Bindings
+    this.childrenChange = this.childrenChange.bind(this);
```
hm, why? :neckbeard: 
all the method of the class already has `this` context, where `this` is instance of the class

dgrammatiko replied:
You can try to remove the bindings and you'll see things will break...

Fedik replied:
nope, not every method, 
it can be need only for `this.getMeter` that used in `add(remove)EventListener` (that changes context)
all else works fin :wink:

Fedik replied:
well, but can stay as is, I just asked, thought maybe I missed something

dgrammatiko replied:
Isn't the same for the fn `childrenChange` as in the mutation observer the context is changed as well due to the arrow function ?
Probably I overdid it (2/4 needs binding) but I want to be safe than sorry...

Fedik replied:
`childrenChange` works because you call it inside arrow function, instead of use it directly as `getMeter`.

Arrow function it is not only  `short hand`, it also has different behavior than generic function,
one of this behavior, is the arrow function do not have `this` and so does not change it:

> An arrow function expression has a shorter syntax than a function expression and does not have its own `this`, `arguments`, `super`, or `new.target`.

There more detailed https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#No_separate_this

PhilETaylor commented:
What about the password box in the installation process? these need doing too :)

C-Lodder commented on build/media/webcomponents/scss/field-password/field-password.scss:
```diff
@@ -0,0 +1,80 @@
+    width: 100%;
+}
+
+joomla-field-password:after {
```
`:after` >> `::after`

dgrammatiko replied:
Maybe no, according to MDN: https://developer.mozilla.org/en-US/docs/Web/CSS/::after
![screenshot 2018-12-03 at 21 03 17](https://user-images.githubusercontent.com/3889375/49398526-f2e61c80-f73e-11e8-8810-274faca77408.png)

C-Lodder replied:
`after` is a pseudo **element**, **NOT** a pseudo class such as `:hover`. The single colon was only used in CSS2 and update for CSS3 to a double colon.

C-Lodder replied:
And single colons for pseudo elements are only used these days to support IE8

dgrammatiko replied:
Ok, my bad. But what is wrong with IE8 most Joomla users (and devs) still use it... 😂

C-Lodder commented on build/media/webcomponents/scss/field-password/field-password.scss:
```diff
@@ -0,0 +1,80 @@
+    line-height: 1.5;
+}
+
+joomla-field-password > .joomla-field-password__show:after {
```
`:after` >> `::after`

dgrammatiko replied:
done

C-Lodder commented on build/media/webcomponents/scss/field-password/field-password.scss:
```diff
@@ -0,0 +1,80 @@
+    background: #3073bb url("data:image/svg+xml;charset=utf8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 576 512'%3E%3Cpath fill='rgb(255,255,255)' d='M569.354 231.631C512.97 135.949 407.81 72 288 72 168.14 72 63.004 135.994 6.646 231.631a47.999 47.999 0 0 0 0 48.739C63.031 376.051 168.19 440 288 440c119.86 0 224.996-63.994 281.354-159.631a47.997 47.997 0 0 0 0-48.738zM288 392c-102.556 ...
+}
+
+joomla-field-password > .joomla-field-password__hide:after {
```
`:after` >> `::after`

C-Lodder commented on build/media/webcomponents/scss/field-password/field-password.scss:
```diff
@@ -0,0 +1,80 @@
+    width: 1px;
+    height: 1px;
+    padding: 0;
+    margin: -1px;
```
Don't use `margin: -1px;` as it can sometimes cause a horizontal scrollbar

C-Lodder commented on build/media/webcomponents/scss/field-password/field-password.scss:
```diff
@@ -0,0 +1,80 @@
+    overflow: hidden;
+    clip: rect(0,0,0,0);
+    border: 0;
+}
```
Add `white-space: nowrap;` too

dgrammatiko replied:
done

C-Lodder commented on build/media/webcomponents/scss/field-password/field-password.scss:
```diff
@@ -0,0 +1,80 @@
+
+joomla-field-password > button {
+    border: none;
+    -webkit-appearance: none;
```
Just need `appearance: none;`. Autoprefixer will deal with the vendor prefixes.

dgrammatiko replied:
done

Hackwar commented:
@dgrammatiko Thank you for this work. I think this is a great addition and I welcome this feature in the registration form. However I disagree with using this in any form of login screen. My main issue is the button to display the cleartext password. I see this as a security issue, which can easily disclose your password to others. I see several scenarios:

* People standing behind you unnoticed
* A person sitting at your PC and asking you to type in your password and then you handing over the controls to that other person again
* In a teamviewer session where you give control to another person

In all these cases, this feature might disclose your password to someone else and thus I would like to see this not included in the logins, at least not by default.

dgrammatiko commented:
@Hackwar actually the password show/hide button should be a progressive enhancement that must be driven from the xml with an attribute line reveal="true". BTW I just checked and half of the J3.x options for password field are not even mentioned in the documentation: https://docs.joomla.org/Password_form_field_type

All and all, I'll make the changes so that the button is xml driven and then I guess people have to decide what the defaults should be. I'm providing the functionality here, I'm not the one deciding how, when or where this should be used...

brianteeman commented:
@hackwar Time to rewrite the web then - it is a common feature

Fedik commented:
>  it is a common feature

originally it was a mobile oriented feature, because limitation of mobile inputs, 
later it was spread everywhere :)

dgrammatiko commented:
> originally it was a mobile oriented feature

mobile IS the internet (>70% of traffic is from mobile devices)

brianteeman commented:
if its not going to be enabled then there really is no point in it

mbabker commented:
> if its not going to be enabled then there really is no point in it

Just look at that wildly successful password strength meter on the field.  Oh, wait...

dgrammatiko commented:
> Just look at that wildly successful password strength meter on the field.

FWIW this custom element is also taking care of that successful option 😉

Hackwar commented:
I'm not saying that we shouldn't use it at all. The registration is a good place where this can be demonstrated and thus people would know it, but to me this shouldn't be something that should be enabled by default in the login form.

dgrammatiko commented:
@Hackwar login page is overridable as the rest of the Joomla, so creating an override (or even Joomla provide some sensible override files) is way to easy for anyone creating a Joomla site. And my point is if you're not able to create an override you shouldn't use Joomla, use wix or whatever because the idea of having one more switch for the field is out of the range of ridiculousness.. 
At some point people need to understand the Joomla will never be able to compete to wix without a proper GUI override creator...

All and all, as I said before this PR is just providing the fumctionality. Where, when or how this is used is another issue (I'm just replacing existing fields with buttons, not introducing them here 😉). So please raise another issue for that, this is not the place for this discussion...

astridx commented:
I applied this branch (git fetch origin pull/23452/head:4.0-dev-CE-password-field) and I run composer install and npm install.

It looks not always like in the pictures in this PR but all elements worked correctly. 

## Backend login form

![test administrator login](https://user-images.githubusercontent.com/9974686/49689949-366acd00-fb29-11e8-826d-71c76e3cd45f.png)



```
<joomla-field-password reveal="true">
                                <input name="passwd" type="password" id="mod-login-password" autocomplete="password" aria-labelledby="mod-login-password-lbl" required>
                        </joomla-field-password>

```

## Module 

![home](https://user-images.githubusercontent.com/9974686/49689957-569a8c00-fb29-11e8-8c37-594460d0452e.png)


```
<joomla-field-password reveal="true">
                                                <input id="modlgn-passwd" type="password" name="password" autocomplete="password" placeholder="Password">
                                        </joomla-field-password>
```


##User edit page

![users edit profile test administration](https://user-images.githubusercontent.com/9974686/49689960-6d40e300-fb29-11e8-9e67-b583fab0e17a.png)


```
<joomla-field-password reveal="true" min-length="4"     >
        <input  autocomplete="password" class="validate-password-strength"   size="30" maxlength="99"   name="jform[password]" type="password" id="jform_password" value="" aria-labeledby="jform_password-lbl">
</joomla-field-password>            
```

astridx commented:
I have tested this item :white_check_mark: successfully on 12f4aaf55f4e2ee5c1ac897573c0c328bd46344e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23215">issues.joomla.org/tracker/joomla-cms/23215</a>.</sub>

astridx commented:
I have tested this item :white_check_mark: successfully on 12f4aaf55f4e2ee5c1ac897573c0c328bd46344e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23215">issues.joomla.org/tracker/joomla-cms/23215</a>.</sub>

dgrammatiko committed:
update

dgrammatiko committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into 4.0-dev-CE-password-field

* '4.0-dev' of github.com:joomla/joomla-cms: (239 commits)
  Fixed inconsistent use of plurals Ref #22791 (#23515)
  [4.0] Adding filter() and postProcessing() to FormField (#12414)
  Csp doc block change. (#23579)
  Fix missing use statement (#23564)
  Move admin menu JS code to es6 (#23155)
  [4.0] Remove show_associations field of association is not enabled (#23501)
  [4.0] Add json to source_formats of template editor (#23470)
  [4.0] [com_menus] Fix "Too few arguments to function (#23535)
  Add $data to event_before_save in tags (#23531)
  [4.0] Fix Legacy Model loading. Fixes #23517 (#23525)
  Update en-GB.com_templates.ini (#23523)
  Reverse order of template params for green/red. Fixes #23519 (#23521)
  Finder Index: Check access to plugins (#23509)
  Update draggable.js (#23487)
  Undefined class \Joomla\CMS\Filesystem\AKFactory (#23506)
  [4.0] Not only set the redirect but actually return (#23498)
  fix notice in syndicate.xml (#23502)
  Fix codestyle
  [4.0] Missing files in template manifests (#23484)
  [4.0] WebAsset: split WebAssetRegistry to Registry and Manager (#23463)
  ...

# Conflicts:
#	modules/mod_login/tmpl/default.php
#	package-lock.json

dgrammatiko committed:
pathe the installation as well

dgrammatiko commented:
So I've changed all the wrong SCSS parts and also updated the installation:
![screenshot 2019-01-20 at 13 01 28](https://user-images.githubusercontent.com/3889375/51438979-f8c69b80-1cb3-11e9-878c-f56f1bc8fc8a.png)
![screenshot 2019-01-20 at 13 01 36](https://user-images.githubusercontent.com/3889375/51438980-fcf2b900-1cb3-11e9-802b-b7401b98edae.png)
![screenshot 2019-01-20 at 13 01 44](https://user-images.githubusercontent.com/3889375/51438981-00864000-1cb4-11e9-9e04-85ca964d9fc3.png)


@wilsonge can you review/merge this?

dgrammatiko committed:
missed one

dgrammatiko commented:
@coolcat-creations I just realise that the backend login page and the installation are not quite in sync in terms of design. Can you decide what should be there and align those?

Also another thing that came up with this PR: as @C-Lodder and @ciar4n were doing the templates and there was a question mark on the forms markup the decision was to keep the old way, make the transaction easier (less code to tweak). The problem is that inputs **SHOULD** always be 100% and the container div should set a specific width according to the breakpoint or the Grid Layout. Having each and every input set to a specific width is bad. You (I mean the team that is dealing with the templates) should change this. We're not in 2010 and we shouldn't break things that are naturally responsive...

coolcat-creations commented:
@dgrammatiko we work on something that adresses the issue of design inconsistences, @Hackwar can you look into this one if we can merge some of our code with this work?

dgrammatiko committed:
Merge branch '4.0-dev' into 4.0-dev-CE-password-field

dgrammatiko commented:
Closing here, I guess ES5 is pretty ok for Joomla

wilsonge commented:
As stated to you on private channels and for the record I’m not merging new custom elements until the existing ones stabilise which is why this has sat there (as has been the case since I first told you this last October/November)

dgrammatiko commented:
@wilsonge whatever, this was my third and last attempt for this
